src/iter/mod.rs

```diff
@@ -825,6 +1025,18 @@ pub trait ParallelIterator: Sized + Send {
 
     /// Creates an iterator over the `Some` items of this iterator, halting
     /// as soon as any `None` is found.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use rayon::prelude::*;
+    ///
+    /// let a = [Some(1), Some(2), Some(4), None, Some(6), Some(8)];
+    ///
+    /// let sum: usize = a.par_iter().cloned().while_some().sum::<usize>();
+    ///
+    /// assert_eq!(sum, 7);
```

Comment:
I think this test is getting lucky too, but perhaps we need to clarify the documented behavior.  It halts as soon as a `None` is found *anywhere*, in arbitrary parallel order.  You can't assume that it will see all of the sequentially-prior `Some` values before finding a `None`, and it may also visit `Some` values that are sequentially after a `None`.  In the extreme, this `while_some` could produce nothing if it hit the `None` first, or it may even produce all of the `Some` items if it hit the `None` last!

Reply:
So what kind of example can I use here ? It's some sort of random that occurs here...

Reply:
Perhaps `while_some().count() < a.len()` is a better example?  It's somewhat random which `Some` elements will make it through, but we can guarantee that it will be less than the total vector length.

You can make the same assertion about length with `filter_map` though - I'm not sure how to clearly demonstrate that `while_some` short-circuits.  Maybe we could show an infinite source iterator getting stopped, like `rayon::iter::repeat(None).while_some().count() == 0`.  Or for another example, [`check_while_some`](https://github.com/rayon-rs/rayon/blob/e16ae7779fd67bcb971cfd4d6a9f03a8df16b626/src/iter/test.rs#L1253-L1273) uses an atomic to count how many values were produced behind the scenes.

As a practical use-case, `FromParallelIterator` [for `Option`](https://github.com/rayon-rs/rayon/blob/e16ae7779fd67bcb971cfd4d6a9f03a8df16b626/src/option.rs#L156) and [for `Result`](https://github.com/rayon-rs/rayon/blob/e16ae7779fd67bcb971cfd4d6a9f03a8df16b626/src/result.rs#L105) each use `while_some()` to short-circuit their collection when they see a `None` or `Err`.

Reply:
Ok, so I will use the test code, it is clear and short.

Thank you for the explaination, really usefull and clear too.
Powerfull `while_some` method !